runtime/src/iree/hal/drivers/hip/hip_multi_queue_command_buffer.c

```diff
@@ -0,0 +1,345 @@
+    iree_hal_queue_affinity_t queue_affinity,
+    const iree_hal_hip_dynamic_symbols_t* hip_symbols,
+    const iree_hal_hip_device_topology_t* topology,
+    iree_host_size_t binding_capacity, iree_allocator_t host_allocator,
+    iree_hal_command_buffer_t** out_command_buffer) {
+  IREE_ASSERT_ARGUMENT(out_command_buffer);
+  *out_command_buffer = NULL;
+
+  if (iree_math_count_ones_u64(queue_affinity) != command_buffer_count) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "expected one command buffer per enabled queue.");
+  }
+
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_hip_multi_queue_command_buffer_t* command_buffer = NULL;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_allocator_malloc(
+              host_allocator,
+              sizeof(*command_buffer) +
+                  command_buffer_count * sizeof(iree_hal_command_buffer_t*) +
+                  iree_hal_command_buffer_validation_state_size(
+                      mode, binding_capacity),
+              (void**)&command_buffer));
+  iree_hal_command_buffer_initialize(
+      device_allocator, mode, command_categories, queue_affinity,
+      binding_capacity,
+      (uint8_t*)command_buffer + sizeof(*command_buffer) +
+          command_buffer_count * sizeof(iree_hal_command_buffer_t*),
+      &iree_hal_hip_multi_queue_command_buffer_vtable, &command_buffer->base);
+  memcpy(command_buffer->child_buffers, in_command_buffers,
+         sizeof(iree_hal_command_buffer_t*) * command_buffer_count);
+
+  command_buffer->command_buffer_count = command_buffer_count;
+  command_buffer->topology = topology;
+  command_buffer->hip_symbols = hip_symbols;
+  *out_command_buffer = (iree_hal_command_buffer_t*)command_buffer;
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static void iree_hal_hip_multi_queue_command_buffer_destroy(
+    iree_hal_command_buffer_t* base_command_buffer) {
+  iree_hal_hip_multi_queue_command_buffer_t* command_buffer =
+      iree_hal_hip_multi_queue_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+  for (uint16_t i = 0; i < command_buffer->command_buffer_count; ++i) {
+    if (command_buffer->child_buffers[i]) {
+      iree_hal_resource_release(command_buffer->child_buffers[i]);
+    }
+  }
+  IREE_TRACE_ZONE_END(z0);
+}
+
+IREE_API_EXPORT bool iree_hal_hip_multi_queue_command_buffer_isa(
+    iree_hal_command_buffer_t* command_buffer) {
+  return iree_hal_resource_is(&command_buffer->resource,
+                              &iree_hal_hip_multi_queue_command_buffer_vtable);
+}
+
+#define CALL_COMMAND(status, command)                                  \
+  iree_hal_queue_affinity_t a = command_buffer->base.queue_affinity;   \
+  uint64_t num = 0;                                                    \
+  uint64_t idx = 0;                                                    \
+  while (a && IREE_LIKELY(iree_status_is_ok(status))) {                \
```

Comment:
hot/cold annotation is already in iree_status_is_ok
```suggestion
  while (a && iree_status_is_ok(status)) {                \
```